Improve common search controller for backward compatibility

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Introduce ControllerResponseBuilder so that `CommonController::searchAction` works on both type of grids (with or without `filterId`) To identify which kind of definition is being used a new `IdentifiableGridDefinitionFactoryInterface` has been introduced along with its matching abstract class `AbstractIdentifiableGridDefinitionFactory` for easier use<br>Also fix the Files filtering
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18218 and #18271
| How to test?  | Check that the search feature is fixed on Employee and Category grid. As some grid have been modified like in this PR https://github.com/PrestaShop/PrestaShop/pull/18130 it might be safer to check those lists<br>- addresses<br>- attachments<br>- orders,<br>- attributes <br>- attribute groups (`sell/catalog/attribute-groups` view action opens attributes list),<br>- customer addresses,<br>- order messages,<br>- merchandise returns,<br>- suppliers ( this was migrated already in 1.7.6, but was still hidden)<br>- CMS page category<br>- tax rules

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/18228)
<!-- Reviewable:end -->

jolelievre committed:
Introduce ControllerResponseBuilder so that CommonController::searchAction works on both type of grids

Progi1984 commented on src/PrestaShopBundle/Service/Grid/ControllerResponseBuilder.php:
```diff
@@ -0,0 +1,105 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
```
declare(strict_types=1);

matks commented on src/PrestaShopBundle/Controller/Admin/CommonController.php:
```diff
@@ -357,15 +359,30 @@ public function searchGridAction(
-        $responseBuilder = $this->get('prestashop.bundle.grid.response_builder');
         /** @var GridDefinitionFactoryInterface $definitionFactory */
         $definitionFactory = $this->get($gridDefinitionFactoryServiceId);
+        $reflectionClass = new ReflectionClass($definitionFactory);
```
I have always been told to never use Reflection in production 😮(ping @PierreRambaud any opinion on this ?)

jolelievre replied:
I'm not a big fan either but I only see two alternatives:
- a try/catch block witch would catch a very generic `FatalThrowaleError` (if ti works)
- adding an interface to all grid definitions which manage the GRID_ID which would allow to use `instanceof`

matks replied:
I ❤️ the interface solution

PierreRambaud replied:
`Interface` FTW

jolelievre replied:
The only issue about the interface solution is that it brings more complexity for module developers, we will have to update the doc accordingly

And I'm afraid it could create a BC break 🤔 as I'm not sure exactly what the interface should contain:
`public function getFilterId(): string` maybe?

PierreRambaud replied:
This complexity already exist in PrestaShop, and if you want to help developers feel free to add an abstract class in addition of the Interface.

matks replied:
No BC break I think 🤔
- create new Interface `IdentifiedGridDefinitionFactoryInterface` that extends existing `GridDefinitionFactoryInterface`
- apply to all existing grids in Core
- check if given definition is instance of `IdentifiedGridDefinitionFactoryInterface` to use the ID

jolelievre committed:
Introduce IdentifiableGridDefinitionFactoryInterface in order to detect what kind of filtering and search needs to be used

jolelievre committed:
Update all grid definition that need to be identifiable and use common controller

jolelievre commented:
@PierreRambaud @matks it should be good, I added the interface and an abstract class Which makes it simpler to update the definition (only change the parent class)
Like in the initial PR https://github.com/PrestaShop/PrestaShop/pull/18130 I only updated the grid that use the common controller, as the remaining grids are planned to be udpated in the next version (in this PR https://github.com/PrestaShop/PrestaShop/pull/18136 I added a comment so that it uses this new parent class)

jolelievre committed:
Fix Files list searching feature

PierreRambaud commented on src/Core/Grid/Definition/Factory/AbstractIdentifiableGridDefinitionFactory.php:
```diff
@@ -0,0 +1,44 @@
+     *
+     * Grid definition already has an internal id used or hooks, that can also be used as an identifier for filters
+     */
+    public function getGridId(): string
+    {
+        return $this->getId();
+    }
```
Still thinking it's confusing, we are already in a Grid context, the getId should be the getGridId. Should be renamed getFilterId

jolelievre committed:
Rename interfaces and method or filters

jolelievre commented:
@PierreRambaud you're right I modified the interface and methods, I'm not 100% satisfied by it but I could't find better naming.
So the interface is `FilterableGridDefinitionFactoryInterface` and the method `getFilterId`

SD1982 commented:
The filters seems to work correctly ! Thanks @jolelievre

matks commented:
Thank you @jolelievre

Progi1984 commented:
Thank you @jolelievre
